packages/devtools_flutter/lib/src/scaffold.dart

```diff
@@ -82,6 +82,23 @@ class DevToolsScaffoldState extends State<DevToolsScaffold>
     _controller = TabController(length: widget.tabs.length, vsync: this);
   }
 
+  // Pushes tab changes into the navigation history.
+  //
+  // Note that this currently works very well, but it doesn't
+  // integrate with the browser's history yet.
+  void _pushScreenToLocalPageRoute(int newIndex) {
+    final previousTabIndex = _controller.previousIndex;
+    if (newIndex != previousTabIndex) {
+      ModalRoute.of(context).addLocalHistoryEntry(LocalHistoryEntry(
+        onRemove: () {
+          if (widget.tabs.length >= previousTabIndex) {
+            _controller.animateTo(previousTabIndex);
+          }
```

Comment:
where is `_controller.previousIndex` modified after tabs are pushed / popped off the stack of indexes

Reply:
Flutter's TabController manages the previousIndex property on its own: https://api.flutter.dev/flutter/material/TabController/index.html

So when the user selects a tab in the tabbar (or swipes on a touchscreen), the index and previousIndex will be updated.

It's also possible to update with controller.animateTo, or by setting index directly for an instantaneous change.